### DISCUSSION TITLE
생성자, 정적 팩토리 메서드, 빌더 패턴

### DISCUSSION THREAD

**Discussion by Starlight258** [Attica-org/athens-backend#27] [General]
모든 경우에서 답은 아니겠지만, 인자의 개수가 많지 않고 객체 생성을 위한 인자가 자주 변경되지 않는다면 
`빌더 패턴` 대신 `정적 팩토리 메서드`를 사용하는 것으로 코딩 컨벤션을 추가하는 건 어떨까요?

### Reference
https://joon2974.tistory.com/16

@big-cir  @yybmion

**Comment by big-cir** [comment]
그거와 관련해서 고민도 했었는데 정적 팩토리의 정확한 이점을 아직 느끼지 못해 우선 builder로 구현했습니다!
@Starlight258 님은 어떤 점이 명확한 이점으로 느껴지셨나요?

**Reply by Starlight258** [reply]
정적 팩토리 메서드를 사용하면 생성자보다 훨씬 할 수 있는 일이 많아진다고 생각해요
하위 객체를 반환할 수도 있고, 매개변수에 따라 다른 클래스의 객체를 반환할 수도 있어요 (파라미터 검증도 가능해요)
입력받은 인자에 따라 메서드 이름을 다르게 작성할 수도 있고, 싱글톤일 경우 이미 생성되어 있는 객체를 반환할수도 있어요.

빌더 패턴은 인자가 많을 경우에 사용하면 좋다고 생각하지만, 그 이외의 경우에는 정적 팩토리 메서드를 사용하는 것이 이점이 있다고 생각해요.

**Comment by Starlight258** [comment]
@big-cir @yybmion 

처음에는 정적 팩토리 메서드가 무조건 좋은 줄 알았는데, 생각해 볼 점이 있네요!

무분별한 정적 팩토리 메서드 사용은 지양해야할 것 같아요!
### 불필요한 코드 중복 
  - 무분별하게 정적 팩토리 메서드를 사용할 경우 private 생성자 코드 + 정적 팩토리 메서드 코드로 **코드 양이 많아진다.**
  - `생성자`를 통해서도 입력값에 대해 유연하게 객체를 생성할 수 있다.
### 객체 지향과 멀어짐
  - 정적 팩토리 메서드는 `static`이므로 클래스 수준에서 동작한다. 상속과 다형성을 적용하기 어렵다.
### 테스트 하기 어려움
  - static이므로 `mocking`, `stubbing`을 적용하기 어렵다.
  
## 언제 정적 팩토리 메서드를 사용해야할까?
-  생성자 외에 추가 로직이 수행되어야할 경우
- 객체 생성 과정이 복잡할 경우
- 객체 생성 시 의도(또는 의미)를 명시적으로 부여하고 싶은 경우

그 외에 단순히 객체를 생성하거나 유연한 객체 생성이 필요한 경우 생성자를 사용하는 게 좋아보입니다.

**Reply by big-cir** [reply]
저번에 언급해주신 부분 자세하게 설명 해주셨군요!
감사합니다! 👍🏻
